lib/services/app.go

```diff
@@ -158,3 +168,222 @@ func UnmarshalAppServer(data []byte, opts ...MarshalOption) (types.AppServer, er
+// in multiple application resources if there are multiple ports exposed.
+func NewApplicationsFromKubeService(service v1.Service, clusterName string, pi ProtocolChecker) ([]types.Application, error) {
+	var apps types.Apps
+	protocol := ""
+	for k, v := range service.GetAnnotations() {
+		if k == types.DiscoveryProtocolLabel {
+			protocol = v
+		}
+	}
+
+	ports, err := getServicePorts(service)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var appsMu sync.Mutex
+	g := errgroup.Group{}
+	g.SetLimit(5)
+	for _, port := range ports {
+		port := port
+		g.Go(func() error {
+			appURI := getAppURI(getServiceFQDN(service), port, protocol, pi)
+
+			// By default we are looking for http apps, we only create tcp apps if protocol annotation
+			// explicitly specifies it.
+			if strings.HasPrefix(appURI, protoTCP) && protocol != protoTCP {
+				return nil
+			}
+
+			rewriteConfig, err := getAppRewriteConfig(service.GetAnnotations())
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			a, err := types.NewAppV3(types.Metadata{
+				Name:        getAppName(service.GetName(), service.GetNamespace(), clusterName, ""),
+				Description: fmt.Sprintf("Discovered application in Kubernetes cluster %q", clusterName),
+				Labels:      getAppLabels(service.GetLabels(), clusterName),
+			}, types.AppSpecV3{
+				URI: appURI,
+				// Temporary usage to have app's name with port name, in case there are more than one app
+				// created from this service.
+				PublicAddr: getAppName(service.GetName(), service.GetNamespace(), clusterName, port.Name),
+				Rewrite:    rewriteConfig,
+			})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			appsMu.Lock()
+			apps = append(apps, a)
+			appsMu.Unlock()
+			return nil
+		})
+	}
+	err = g.Wait()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// If we ended up with more than one app from the service, we'll have port name in the app name to distinguish them.
+	// We can only do this now because we don't know how many apps there will be until all ports are processed by errgroup.
+	for i := range apps {
+		if len(apps) > 1 {
+			apps[i].SetName(apps[i].(*types.AppV3).Spec.PublicAddr)
+		}
+		apps[i].(*types.AppV3).Spec.PublicAddr = "" // Clear temporary used for name field.
+	}
+
+	return apps, nil
+}
+
+// ProtocolChecker is an interface used to check what protocol uri serves
+type ProtocolChecker interface {
+	CheckProtocol(uri string) string
+}
+
+func getServiceFQDN(s v1.Service) string {
+	host := s.GetName()
+	if s.Spec.Type == v1.ServiceTypeExternalName {
+		host = s.Spec.ExternalName
+	}
+	return fmt.Sprintf("%s.%s.svc.cluster.local", host, s.GetNamespace())
+}
+
+const (
+	protoHTTPS = "https"
+	protoHTTP  = "http"
+	protoTCP   = "tcp"
+)
+
+func getAppURI(serviceFQDN string, port v1.ServicePort, protocolAnnotation string, pc ProtocolChecker) string {
+	constructURI := func(p string) string { return fmt.Sprintf("%s://%s:%d", p, serviceFQDN, port.Port) }
```

Comment:
use the url.Url to build it since it reads better.

Reply:
I don't see much improvement tbh, we end up with almost same `Sprintf()` for host + `url.Url` stuff.